rust/sbp/src/sbp_tools.rs

```diff
@@ -0,0 +1,259 @@
+        Self: Iterator<Item = Result<SBP>> + Sized,
+    {
+        HandleErrorsIter::new(self, |e| {
+            log::warn!("{}", e);
+            true
+        })
+    }
+
+    fn handle_errors<F>(self, on_err: F) -> HandleErrorsIter<Self, F>
+    where
+        Self: Iterator<Item = Result<SBP>> + Sized,
+        F: FnMut(&crate::Error) -> bool,
+    {
+        HandleErrorsIter::new(self, on_err)
+    }
+
+    #[cfg(feature = "swiftnav-rs")]
+    fn with_rover_time(self) -> RoverTimeIter<Self>
+    where
+        Self: Sized,
+    {
+        RoverTimeIter::new(self)
+    }
+}
+
+impl<I> SBPTools for I where I: Iterator + Sized {}
+
+pub struct HandleErrorsIter<I, F>
+where
+    I: Iterator,
+{
+    messages: I,
+    on_err: F,
+    err: Result<()>,
+}
+
+impl<I, F> HandleErrorsIter<I, F>
+where
+    I: Iterator<Item = Result<SBP>>,
+{
+    fn new(messages: I, on_err: F) -> HandleErrorsIter<I, F> {
+        Self {
+            messages,
+            on_err,
+            err: Ok(()),
+        }
+    }
+
+    pub fn check_error(&mut self) -> Result<()> {
+        std::mem::replace(&mut self.err, Ok(()))
+    }
+}
+
+impl<I, F> Iterator for HandleErrorsIter<I, F>
+where
+    I: Iterator<Item = Result<SBP>>,
+    F: FnMut(&crate::Error) -> bool,
+{
+    type Item = SBP;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.messages.next()? {
+            Ok(msg) => Some(msg),
+            Err(e) => {
+                let should_continue = (self.on_err)(&e);
+                self.err = Err(e);
+                if should_continue {
+                    self.next()
+                } else {
+                    None
+                }
+            }
+        }
+    }
+}
+
+#[cfg(feature = "swiftnav-rs")]
+pub struct RoverTimeIter<I: Iterator> {
+    messages: I,
+    clock: Option<GpsTime>,
+}
+
+#[cfg(feature = "swiftnav-rs")]
+impl<I> RoverTimeIter<I>
+where
+    I: Iterator,
+{
+    fn new(messages: I) -> RoverTimeIter<I> {
+        Self {
+            messages,
+            clock: None,
+        }
+    }
+
+    fn update(&mut self, time: MessageTime) -> Option<GpsTime> {
+        match time {
+            MessageTime::Rover(time) => match time {
+                RoverTime::GpsTime(time) => {
+                    self.clock = Some(time);
+                    Some(time)
+                }
+                RoverTime::Tow(tow) => {
+                    if let Some(clock) = self.clock {
+                        // tow came from a `GpsTime` so it must be valid
+                        let time = GpsTime::new(clock.wn(), tow).unwrap();
+                        self.clock = Some(time);
+                        Some(time)
+                    } else {
+                        None
+                    }
+                }
+            },
+            _ => None,
+        }
+    }
+}
+
+#[cfg(feature = "swiftnav-rs")]
+impl<I> Iterator for RoverTimeIter<I>
+where
+    I: Iterator<Item = SBP>,
+{
+    type Item = (SBP, Option<GpsTime>);
```

Comment:
should probably be `(SBP, Option<Result<GpsTime>>)`

Reply:
agree